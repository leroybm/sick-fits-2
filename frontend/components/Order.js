import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import PropTypes from 'prop-types';
import Head from 'next/head';
import DisplayError from './ErrorMessage';
import OrderStyles from './styles/OrderStyles';
import formatMoney from '../lib/formatMoney';

const ORDER_QUERY = gql`
  query ORDER_QUERY($id: ID!) {
    Order(where: { id: $id }) {
      id
      label
      total
      charge
      user {
        id
      }
      items {
        id
        name
        description
        price
        quantity
        photo {
          altText
          image {
            publicUrlTransformed
          }
        }
      }
    }
  }
`;
export default function Order({ id }) {
  const { data, loading, error } = useQuery(ORDER_QUERY, { variables: { id } });

  if (loading) return <p>Loading...</p>;
  if (error) return <DisplayError error={error} />;
  const order = data.Order;

  return (
    <OrderStyles>
      <Head>
        <title>Sick Fits - {order.id}</title>
      </Head>

      <p>
        <span>Order Id:</span>
        <span>{order.id}</span>
      </p>

      <p>
        <span>Charge:</span>
        <span>{order.charge}</span>
      </p>

      <p>
        <span>Order Total:</span>
        <span>{formatMoney(order.total)}</span>
      </p>

      <p>
        <span>ItemCount:</span>
        <span>{order.items.length}</span>
      </p>

      <div className="items">
        {order.items.map((item) => (
          <div className="order-item" key={item.id}>
            <img
              src={item.photo.image.publicUrlTransformed}
              alt={item.photo.altText}
            />
            <div className="item-details">
              <h2>{item.name}</h2>
              <p>Qty: {item.quantity}</p>
              <p>Each: {formatMoney(item.price)}</p>
              <p>Subtotal: {formatMoney(item.price * item.quantity)}</p>
              <p>{item.description}</p>
            </div>
          </div>
        ))}
      </div>
    </OrderStyles>
  );
}

Order.propTypes = {
  id: PropTypes.string,
};
