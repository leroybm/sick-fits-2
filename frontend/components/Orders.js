import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import styled from 'styled-components';
import Link from 'next/link';
import Head from 'next/head';
import DisplayError from './ErrorMessage';
import OrderItemStyles from './styles/OrderItemStyles';
import formatMoney from '../lib/formatMoney';

const USER_ORDER_QUERY = gql`
  query USER_ORDER_QUERY {
    allOrders {
      id
      label
      total
      charge
      user {
        id
      }
      items {
        id
        name
        description
        price
        quantity
        photo {
          altText
          image {
            publicUrlTransformed
          }
        }
      }
    }
  }
`;

const OrderUl = styled.ul`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(350px, 1fr));
  grid-gap: 4rem;
`;

function InlineOrder({ order }) {
  const itemQuantity = order.items.reduce(
    (tally, item) => tally + item.quantity,
    0
  );

  return (
    <OrderItemStyles key={order.id}>
      <Link href={`/order/${order.id}`}>
        <a>
          <div className="order-meta">
            <p>
              {itemQuantity} Item{itemQuantity > 1 ? 's' : ''}
            </p>
            <p>
              {order.items.length} Product{order.items.length === 1 ? '' : 's'}
            </p>
            <p>{formatMoney(order.total)}</p>
          </div>
          <div className="images">
            {order.items.map((item) => (
              <img
                key={item.id}
                src={item.photo.image?.publicUrlTransformed}
                alt={item.photo.altText}
              />
            ))}
          </div>
        </a>
      </Link>
    </OrderItemStyles>
  );
}

export default function Orders() {
  const { data, loading, error } = useQuery(USER_ORDER_QUERY);

  if (loading) return <p>Loading...</p>;
  if (error) return <DisplayError error={error} />;
  const { allOrders } = data;

  return (
    <div>
      <Head>
        <title>Your Orders ({allOrders.length})</title>
      </Head>
      <h2>You have {allOrders.length} orders!</h2>

      <OrderUl>
        {allOrders.map((order) => (
          <InlineOrder order={order} key={order.id} />
        ))}
      </OrderUl>
    </div>
  );
}
