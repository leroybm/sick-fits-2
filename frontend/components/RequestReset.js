import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      code
      message
    }
  }
`;

export default function RequestReset() {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
  });

  const [requestReset, { loading, error, data }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: inputs,
    }
  );

  return (
    <Form
      method="POST"
      onSubmit={async (event) => {
        event.preventDefault();
        await requestReset().catch(console.error);
        resetForm();
      }}
    >
      <h2>Request A Password Reset</h2>

      <DisplayError error={error} />

      <fieldset disabled={loading} aria-busy={loading}>
        {data?.sendUserPasswordResetLink === null && (
          <p>Success! Check your email for a link!</p>
        )}

        <label htmlFor="email">
          Email
          <input
            type="email"
            id="email"
            name="email"
            placeholder="Email"
            value={inputs.email}
            onChange={handleChange}
            autoComplete="email"
          />
        </label>

        <button type="submit">Request Rest!</button>
      </fieldset>
    </Form>
  );
}
