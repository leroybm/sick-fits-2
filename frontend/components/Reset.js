import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';

const RESET_MUTATION = gql`
  mutation RESET_MUTATION(
    $email: String!
    $password: String!
    $token: String!
  ) {
    redeemUserPasswordResetToken(
      email: $email
      password: $password
      token: $token
    ) {
      message
      code
    }
  }
`;

export default function Reset({ token }) {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
    password: '',
  });

  const [requestReset, { loading, error, data }] = useMutation(RESET_MUTATION, {
    variables: { ...inputs, token },
  });

  const resetError = data?.redeemUserPasswordResetToken?.code
    ? data.redeemUserPasswordResetToken
    : null;
  return (
    <Form
      method="POST"
      onSubmit={async (event) => {
        event.preventDefault();
        await requestReset().catch(console.error);
        resetForm();
      }}
    >
      <h2>Reset Your Password</h2>

      <DisplayError error={error || resetError} />

      <fieldset disabled={loading} aria-busy={loading}>
        {data?.redeemUserPasswordResetToken === null && (
          <p>Success! Your password was reset!</p>
        )}

        <label htmlFor="email">
          Email
          <input
            type="email"
            id="email"
            name="email"
            placeholder="Email"
            value={inputs.email}
            onChange={handleChange}
            autoComplete="email"
          />
        </label>

        <label htmlFor="password">
          Password
          <input
            type="password"
            id="password"
            name="password"
            placeholder="Password"
            value={inputs.password}
            onChange={handleChange}
            autoComplete="password"
          />
        </label>

        <button type="submit">Request Rest!</button>
      </fieldset>
    </Form>
  );
}
