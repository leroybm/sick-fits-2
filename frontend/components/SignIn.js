import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';
import { CURRENT_USER_QUERY } from './User';

const AUTHENTICATE_USER_MUTATION = gql`
  mutation AUTHENTICATE_USER_MUTATION($email: String!, $password: String!) {
    authenticateUserWithPassword(email: $email, password: $password) {
      ... on UserAuthenticationWithPasswordSuccess {
        sessionToken
        item {
          id
          name
        }
      }

      ... on UserAuthenticationWithPasswordFailure {
        code
        message
      }
    }
  }
`;

export default function SignIn() {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
    password: '',
  });

  const [signInUser, { loading, error, data }] = useMutation(
    AUTHENTICATE_USER_MUTATION,
    { variables: inputs, refetchQueries: [{ query: CURRENT_USER_QUERY }] }
  );

  const authError =
    data?.authenticateUserWithPassword.__typename ===
    'UserAuthenticationWithPasswordFailure'
      ? data.authenticateUserWithPassword
      : null;
  return (
    <Form
      method="POST"
      onSubmit={async (event) => {
        event.preventDefault();
        await signInUser();
        resetForm();
      }}
    >
      <h2>Sign In Your Account</h2>

      <DisplayError error={error || authError} />

      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="email">
          Email
          <input
            type="email"
            id="email"
            name="email"
            placeholder="Email"
            value={inputs.email}
            onChange={handleChange}
            autoComplete="email"
          />
        </label>

        <label htmlFor="password">
          Password
          <input
            type="password"
            id="password"
            name="password"
            placeholder="Password"
            value={inputs.password}
            onChange={handleChange}
            autoComplete="password"
          />
        </label>

        <button type="submit">Sign In</button>
      </fieldset>
    </Form>
  );
}
