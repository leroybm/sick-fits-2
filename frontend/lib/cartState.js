import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  const [cartOpen, setCartOpen] = useState(false);

  return (
    <LocalStateProvider value={{ cartOpen, setCartOpen }}>
      {children}
    </LocalStateProvider>
  );
}

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
