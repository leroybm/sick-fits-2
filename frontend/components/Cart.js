import styled from 'styled-components';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import CartStyles from './styles/CartStyles';
import { CURRENT_USER_QUERY, useUser } from './User';
import Supreme from './styles/Supreme';
import formatMoney from '../lib/formatMoney';
import calcTotalPrice from '../lib/calcTotalPrice';
import { useCart } from '../lib/cartState';
import CloseButton from './styles/CloseButton';
import RemoveFromCart from './RemoveFromCart';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid var(--lightGrey);
  display: grid;
  grid-template-columns: auto 1fr auto;

  img {
    margin-right: 1rem;
  }

  h3,
  p {
    margin: 0;
  }
`;

function CartItem({ cartItem }) {
  const { product } = cartItem;

  if (!product) return null;

  return (
    <CartItemStyles>
      <img
        width="100"
        src={product.photo.image.publicUrlTransformed}
        alt={product.name}
      />
      <div>
        <h3>{product.name}</h3>
        <p>
          {formatMoney(product.price * cartItem.quantity)}&nbsp;-&nbsp;
          <em>
            {cartItem.quantity} &times; {formatMoney(product.price)}
          </em>
        </p>
      </div>
      <RemoveFromCart id={cartItem.id} />
    </CartItemStyles>
  );
}

export default function Cart() {
  const me = useUser();
  const { cartOpen, setCartOpen } = useCart();

  if (!me) return null;

  return (
    <CartStyles open={cartOpen}>
      <header>
        <Supreme>{me.name}'s Cart</Supreme>
        <CloseButton type="button" onClick={() => setCartOpen(false)}>
          &times;
        </CloseButton>
      </header>

      <ul style={{ overflow: 'auto' }}>
        {me.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>

      <footer>
        <p>{formatMoney(calcTotalPrice(me.cart))}</p>
      </footer>
    </CartStyles>
  );
}
