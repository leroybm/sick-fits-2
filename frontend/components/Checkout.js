import styled from 'styled-components';
import { loadStripe } from '@stripe/stripe-js';
import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from '@stripe/react-stripe-js';
import { useState } from 'react';
import nProgress from 'nprogress';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { useRouter } from 'next/dist/client/router';
import SickButton from './styles/SickButton';
import { useCart } from '../lib/cartState';
import { CURRENT_USER_QUERY } from './User';

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.04);
  border: 1px solid rgba(0, 0, 0, 0.06);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;
`;

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

function CheckoutForm() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();
  const [checkout, { error: graphQLError }] = useMutation(
    CREATE_ORDER_MUTATION,
    { refetchQueries: [{ query: CURRENT_USER_QUERY }] }
  );
  const router = useRouter();
  const { setCartOpen } = useCart();

  async function handleSubmit(event) {
    // Stop the form from submitting and turn the loader on
    event.preventDefault();
    setLoading(true);

    // Start the page transition
    nProgress.start();

    // Create the payment method via stripe (Token comes back here if successful)
    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      card: elements.getElement(CardElement),
    });

    if (nProgress.status < 0.66) {
      nProgress.set(0.66);
    }

    // Handle any errors from stripe
    if (error) {
      setError(error);
      nProgress.done();
      return;
    }

    // Send the token from step 3 to our keystone server via a custom mutation!
    const order = await checkout({ variables: { token: paymentMethod.id } });
    console.log('FINISHED ORDER');
    console.log(order);

    // Change the page to view the order
    router.push({
      pathname: `/order/[id]`,
      query: { id: order.data.checkout.id },
    });

    // Close the cart
    setCartOpen(false);

    // Turn the loader off
    setLoading(false);
    nProgress.done();
  }

  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      {error && <p style={{ fontSize: 12 }}>{error.message}</p>}
      {graphQLError && <p style={{ fontSize: 12 }}>{graphQLError.message}</p>}
      <CardElement />
      <SickButton>Check Out Now</SickButton>
    </CheckoutFormStyles>
  );
}

export default function Checkout() {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}
