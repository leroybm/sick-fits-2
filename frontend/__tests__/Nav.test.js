import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { debug } from 'prettier/doc';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeCartItem, fakeUser } from '../lib/testUtils';
import { CartStateProvider } from '../lib/cartState';
import Nav from '../components/Nav';

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: fakeUser() } },
  },
];

const signedInWithCartItemsMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        authenticatedItem: fakeUser({
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
        }),
      },
    },
  },
];

describe('<Nav />', () => {
  it('should render minimal nav when signed out', () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider mocks={notSignedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );

    expect(container).toMatchSnapshot();
    [
      {
        text: 'Products',
        href: '/products',
      },
      {
        text: 'Sign In',
        href: '/signin',
      },
    ].forEach(({ text, href }) => {
      expect(container).toHaveTextContent(text);
      const link = screen.getByText(text);
      expect(link).toHaveAttribute('href', href);
    });
  });

  it('should render a full nav when signedin', async () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider mocks={signedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByText('Account');

    expect(container).toMatchSnapshot();
    [
      {
        text: 'Products',
        href: '/products',
      },
      {
        text: 'Sell',
        href: '/sell',
      },
      {
        text: 'Orders',
        href: '/orders',
      },
      {
        text: 'Account',
        href: '/account',
      },
    ].forEach(({ text, href }) => {
      expect(container).toHaveTextContent(text);
      const link = screen.getByText(text);
      expect(link).toHaveAttribute('href', href);
    });

    expect(container).toHaveTextContent('Sign Out');
    expect(container).toHaveTextContent('My Cart');
  });

  it('should render the amount of items in the cart', async () => {
    render(
      <CartStateProvider>
        <MockedProvider mocks={signedInWithCartItemsMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByText('Account');
    expect(screen.getByText(9)).toBeInTheDocument();
  });
});
