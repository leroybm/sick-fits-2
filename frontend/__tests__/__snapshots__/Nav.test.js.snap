// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> should render a full nav when signedin 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 jofUXi"
  >
    <a
      href="/products"
    >
      Products
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/account"
    >
      Account
    </a>
    <button
      type="button"
    >
      Sign Out
    </button>
    <button
      type="button"
    >
      My Cart
    </button>
  </ul>
</div>
`;

exports[`<Nav /> should render minimal nav when signed out 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 jofUXi"
  >
    <a
      href="/products"
    >
      Products
    </a>
    <a
      href="/signin"
    >
      Sign In
    </a>
  </ul>
</div>
`;
